src/ngraph/specialize_shapes.hpp

```diff
@@ -0,0 +1,46 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include "ngraph/function.hpp"
+
+namespace ngraph
+{
+    /// \brief Creates a clone of a function, with the shapes of that function's parameters
```

Comment:
The description you have in the PR is excellent. So excellent I think it belongs in the header. It is more "tutorial" but I understood a lot from reading it.